### PR TITLE
l10n_ch*: added QR-IBAN on bank account for Switzerland accounting

### PR DESCRIPTION
**Pull request by rgo-odoo**
Task: <a href="https://www.odoo.com/web?#action=4043&cids=2&id=2307262&menu_id=4720&model=project.task&view_type=form">2307262</a>

### CODE REVIEW

**Review by yvaucher — CHANGES_REQUESTED** [review]
QR-IBAN from invoice.invoice_partner_bank_id.l10n_ch_qr_iban must also be displayed on the report here:

https://github.com/odoo/odoo/blob/13.0/addons/l10n_ch/report/swissqr_report.xml#L103

**Comment by fclementic2c** [comment]
> QR-IBAN from invoice.invoice_partner_bank_id.l10n_ch_qr_iban must also be displayed on the report here:

Yes the report is not valid at the moment (just tested it with a customer)

**Comment by yvaucher** [comment]
ping @jco-odoo

**Comment by jco-odoo** [comment]
@yvaucher I am on it today.  (if you have any suggestions...)

**Comment by yvaucher** [comment]
@jco-odoo great, it's just to notice you on it. Only the visible part of the report is not ok. The QR code contains already the good information. It's just  a matter of syncing the content of the QR-Code and the informations displayed on the report.

We need to support either a customer that reads the report and copy it by hand or a customer that scan the QR code.

**Comment by jco-odoo** [comment]
@yvaucher @fclementic2c I still want to do some clean up/review, but this changes the report already.  Do QR-IBANs only work with IBAN bank accounts?  Because I would make the QR-IBAN field visible on the bank account when the subscription fields are visible as well.  

@tsb-odoo

**Comment by fclementic2c** [comment]
Improvement suggestion @jco-odoo : Please add the QR IBAN field on the view : account.online.sync.res.partner.bank.setup.form from model : account.setup.bank.manual.config
Go to Accounting / Configuration / Add a bank account -> Create it

screenshot-fclementic2c-swiss-accounting odoo com-2020 09 01-11_37_35

**Comment by yvaucher** [comment]
@jco-odoo I understand that you ask if there are any other kind of bank account in Switzerland than IBAN that could use a QR-IBAN.

The answer is no.

All QR-IBAN have a bank account with an IBAN that lies behind, I see no exception.
When you are provided a QR-IBAN from a supplier you don't need to know the IBAN behind it.

When you generate QR-Bill, you know both your IBAN and the linked QR-IBAN.

**Comment by jco-odoo** [comment]
@yvaucher I could add the condition again for only showing the field when it is IBAN, but I think the most important is to show it when it is about a company bank account like the subscription numbers.

**Comment by yvaucher** [comment]
I agree with you, that field is intended for the company bank accounts. Hiding it based on `acc_type` could confuse the users.

When you go to setup your company bank and see the field afterward I would guess that a confused user would:

1. Create new bank account
2. Fill QR-IBAN in IBAN field
3. See the QR-IBAN that just appeared :confused: 
4. Set the QR-IBAN again but in the other fiield this time and curse the program or it's developpers.

**Comment by jco-odoo** [comment]
@oco-odoo Can you have a quick look as well?

File: addons/l10n_ch_qr/models/res_bank.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            self._validate_ch_qr_iban(vals['l10n_ch_qr_iban'])
+            vals['l10n_ch_qr_iban'] = pretty_iban(normalize_iban(vals['l10n_ch_qr_iban']))
+        return super().write(vals)
+
+    def _is_qr_iban(self):
+        res = super()._is_qr_iban()
+        # If the acc_number is valid QR-IBAN, we will use acc_number to build SWISS code URL.
+        if not res and self.l10n_ch_qr_iban:
+            res = True
+        return res
```
Comment (oco-odoo) [inline addons/l10n_ch_qr/models/res_bank.py:44]:
```suggestion
        return super()._is_qr_iban() or self.l10n_ch_qr_iban
```

**Review by oco-odoo — CHANGES_REQUESTED** [review]
Also, I assume the idea is to merge this new module with l10n_ch when the forward-port reaches master?

File: addons/l10n_ch_qr/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+Swiss localization
+==================
+Added a QR-IBAN field on bank account.
+If this field is empty, but the bank account number itself is a valid QR-IBAN number, it will still be used as QR-IBAN as before.  
+But if you fill in the new QR-IBAN field, that one will be used as the QR-IBAN.  This should help for reconciliation as 
+on the bank statements, the old IBAN code is still used.  
+    """,
+    'version': '1.0',
+    'author': 'Odoo S.A',
+    'category': 'Accounting/Localizations',
```
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_ch_qr/__manifest__.py:15]: Indeed.  Apparently there was not much respect for it at the time.

File: addons/l10n_ch_qr/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_ch_qr/__manifest__.py:1]: qr_iban tends to be a bridge module for qr and iban, so I suggest l10n_ch_qriban

**Comment by fclementic2c** [comment]
@jco-odoo @oco-odoo Any chance to merge this ? it is a problem on our side. Thanks

**Comment by jco-odoo** [comment]
@fclementic2c To me, this is ok to merge.  I just give a bit of time for the functional check.

**Comment by jco-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**